### ISSUE TITLE
Listview and InvalidateCacheEntryAsync

### ISSUE THREAD

**Issue by tallmanBS** [luberda-molinet/FFImageLoading#835]
Hello,
Love the product.  I'm stuck understanding maybe the key, or maybe InvalidateCacheEntryAsync.  
mvvm Firstpage shows image (inside listview)
.  Calls mvvm Secondpage to edit/change image (same URL).  On second page I invalidateChacheEntryAsync on the image, and it changes it and shows new image.  WHen I return to first page it still shows old image.  I've also InvalidateCacheEntryAsync on First page but it never changes the image.  I can't figure out why or how.  Even if I close app and reopen, still showing cached version of image.  I'm using the Image URL (full name including https://) as the key.  

How can I diagnose this?  I've tried turning   ImageService.Instance.Config.VerboseLogging = true;
But apparantly I can't do anything right because I can't get the debug messages to show anywhere.

Help appreciated.
Thanks.
rent

**Comment by daniel-luberda** [comment]
> ImageService.Instance.Config.VerboseLogging

Don't do that, instead: https://github.com/luberda-molinet/FFImageLoading/blob/master/samples/ImageLoading.Forms.Sample/Droid/MainActivity.cs#L29-L37


> WHen I return to first page it still shows old image. I've also InvalidateCacheEntryAsync on First page but it never changes the image. I can't figure out why or how. Even if I close app and reopen, still showing cached version of image. I'm using the Image URL (full name including https://) as the key.

It won't automatically refresh all images that were previously shown. That's why, you should call `ReloadImage` method to refresh it.

**Comment by tallmanBS** [comment]
Thanks for debug, got that working.  So is the cache per image and view then?

I'm struggling to reload this because I'm on mvvm, and it's within a listview, and I don't have reload in the model.  So I'm believing I need to call a function on the page, and loop through the listview looking for that image/row, and reload?  

FYI, just donated, thanks for your software and your hard work supporting it.
Brent

**Comment by daniel-luberda** [comment]
> So is the cache per image and view then?

No, it isn't, there is only one cache used for all views. But cache is used only for image loading operations. After you load it, it's up to dev if it should stay or change. 

> I'm struggling to reload this because I'm on mvvm, and it's within a listview, and I don't have reload in the model. 

It should be easy to do with a messaging system. You just subscribe for an "RefreshImages" message, and call _image.ReloadImage() when you receive it. Do you use Xamarin.Forms? Take a look at: 

- https://developer.xamarin.com/guides/xamarin-forms/application-fundamentals/messaging-center/
- https://github.com/runceel/AsyncMessenger

You can even subscribe for reloading specific cache key. That could help: 

```C#
public class CustomCacheImage : CachedImage
{
    public CustomCacheImage()
    {
        Success += (sender, e) => { CacheKey = e.ImageInformation.CacheKey; };

        MessagingCenter.Subscribe<SenderType, string>(this, "ReloadImage", (sender, key) =>
        {
            if (key == CacheKey)
                ReloadImage();
        });
    }

    public string CacheKey { get; private set; }
}
```

Then use call reload like: 

```C#
MessagingCenter.Send(this, "ReloadImage", "some key to reload");
```

> FYI, just donated, thanks for your software and your hard work supporting it.

Thank you! I really appreciate it, it's always nice to hear that it's useful for someone.

**Comment by tallmanBS** [comment]
Thanks again for the insight and code suggestion.  I've avoided messaging center cause it feels like a workaround hack, albiet maybe easier in this case.
I tried to be clever and create a bindable property to do the reload so I could just bind that to my object.

`    public class CustomCachedImage : CachedImage
    {
        public bool ReloadIt
        {
            get { return (bool)GetValue(ReloadItProperty); }
            set { SetValue(ReloadItProperty, value); }
        }

        public static readonly BindableProperty ReloadItProperty =
                   BindableProperty.Create("BindableReloadIt", typeof(bool), typeof(CachedImage),
                       true, BindingMode.TwoWay, propertyChanged: OnReloadIt);

        private static void OnReloadIt(BindableObject bindable, object oldvalue, object newvalue)
        {
            var item = bindable as CustomCachedImage;

            if (item != null && item.Source != null && (bool)newvalue == true && (bool)oldvalue == false)
            {
                    item.ReloadImage();
                    item.ReloadIt = false;
            }
        }
    }
`

And it does trigger the reload, but it's getting called I think at close of window with an object null reference on image.ReloadImage().   So does this seem like a good idea to do, and if so how can I add an if condition so it doesn't attempt to reload outside of scope.  The item in this case seems fully populated, but maybe the Android view (InternalReloadImage) isn't there.
Brent

**Comment by daniel-luberda** [comment]
Your solution is also a good one. Did you try to wrap the code with `try {} catch {}` block? `CachedImage` view can be in state without it's renderer attached.

**Comment by daniel-luberda** [comment]
You could also use Behaviors (https://developer.xamarin.com/guides/xamarin-forms/application-fundamentals/behaviors/creating/) - and handle PropertyChanged event - they're attached / detached automatically.

**Comment by tallmanBS** [comment]
I wrapped to allow to continue.  Still not updating the image.  I know reload is getting executed.  Do I need to trigger somehow propertychanged on the listview or entire cacheimage?  Although I say that, I don't know how to do it.  I'll probably try your solution with messaging as I'm just not knowledgable enough on this.
Brent

**Comment by tallmanBS** [comment]
Still stuck.  When I attempt to use the messaging service, or my custom cache with binding, I'm using at a point where the renderer must not be loaded, because when I hit
                    ReloadImage();
I get object null reference.  The Try catch can catch it, but it never executes so it's not reloading the image. 

My previous page (where image is changed) actually has a handle to the first page, and calls the first page as the second is popped, so maybe the view is not attached at that point.
If behaviors is the solution, do I write the behavior for the listview on attached?  THere is no onattached for the imagedcache (at least I don't think so), so I'm not sure what behavior control it should override.

**Comment by daniel-luberda** [comment]
```C#
public class ReloadBehavior : Behavior<CachedImage>
{
    CachedImage _view;

    protected override void OnAttachedTo (CachedImage bindable)
    {
        base.OnAttachedTo (bindable);
        _view = bindable;
        bindable.PropertyChanged += OnPropertyChanged;
    }

    protected override void OnDetachingFrom (CachedImage bindable)
    {
        base.OnDetachingFrom (bindable);
        _view = null;
        bindable.PropertyChanged -= OnPropertyChanged;
    }

    OnPropertyChanged(string propertyNAme)
    {
        if (propertyName == "ReloadIt")
            _view?.ReloadImage();
    }
}
```

Then use it: 

```C#
  public class CustomCachedImage : CachedImage
   {
       public CustomCachedImage()
       {
           Behaviors.Add(new ReloadBehavior());
       }
       
       public bool ReloadIt
       {
           get { return (bool)GetValue(ReloadItProperty); }
           set { SetValue(ReloadItProperty, value); }
       }

       public static readonly BindableProperty ReloadItProperty =
                  BindableProperty.Create("BindableReloadIt", typeof(bool), typeof(CachedImage), default(bool));
   }
```

**Comment by tallmanBS** [comment]
I received errors on your bindable code.  	CS0123	No overload for 'OnPropertyChanged' matches delegate 'PropertyChangedEventHandler'	
I changed to:
```
        void OnPropertyChanged(object sender, PropertyChangedEventArgs e)
        {
            if (e.PropertyName == "ReloadIt")
                _view?.ReloadImage();
        }
```
However, this never gets triggered when propertyname='ReloadIt'.  Wouldn't the onattachedto only get this to trigger when the view is first appearing?  Rather than anytime I want it to change when returning from another page?

**Comment by daniel-luberda** [comment]
> Wouldn't the onattachedto only get this to trigger when the view is first appearing?

No as you subscribe to `PropertyChanged` event there, so you listen to every property changes after that. More info is here: https://developer.xamarin.com/guides/xamarin-forms/application-fundamentals/behaviors/

Could you make a repo project?

**Comment by daniel-luberda** [comment]
@tallmanBS Did you solve it?

**Comment by tallmanBS** [comment]
I couldn't come up with a way for you to test as I would need to give access to our restful service to push the image in a child page.  I think I came back and played with it and think I solved, but adding PropertyChanged to my class that was the source of the listview.  I didn't know I needed to do that (using Fody and [ImplementPropertyChanged]) within my classes for it to update the listview items correctly when changed. I actually read a post from on on different forum oddly enough. When I did that, it the image now changes to the new image.  I do reload the list and the image changes.  Hope that's what was expected.
So I think this is good and resolved for me.  Thanks for checking and following up.
Brent
